### PR TITLE
Time for version 3.0

### PR DESCRIPTION
**Pull request by cristianoc72**
- Bump to PHP 8.0+
- Update dependencies
- Introduce the new php functionalities, in particular union types and new string manipulation functions.
- Refactor Json library to use native php `JsonException`. Basically, it forces `json_encode` and `json_decode` to raise `JsonException` instead of old-style errors
- Fix all Psalm issue by annotating the code where psalm can't infer types. Probably, we could remove some annotations with future versions of Psalm. Now Psalm is green at level 1

### CODE REVIEW

File: src/file/FileOperationTrait.php [hunk 1]
```diff
@@ -15,30 +15,23 @@
...
-	 * @psalm-suppress UnsafeInstantiation
+	 * @return self
 	 */
-	public static function create(Stringable | string $pathname): static {
-		return new static($pathname);
+	public static function create(string|Stringable $pathname): self {
+		return new self($pathname);
+	}
+
+	final public function __construct(string|Stringable $filename) {
```
Comment (gossi) [inline src/file/FileOperationTrait.php]: Dunno, if this is a good idea to move the constructor to a trait... 🤔

File: tests/json/JsonTest.php [hunk 2]
```diff
@@ -104,31 +104,31 @@ public function testDecodeBrokenStackDepth(): void {
...
 
 		if (version_compare(PHP_VERSION, '5.5.0') >= 0) {
-			$this->assertEquals(JSON_ERROR_RECURSION, Json::ERROR_RECURSION);
-			$this->assertEquals(JSON_ERROR_INF_OR_NAN, Json::ERROR_INF_OR_NAN);
-			$this->assertEquals(JSON_ERROR_UNSUPPORTED_TYPE, Json::ERROR_UNSUPPORTED_TYPE);
-			$this->assertEquals(JSON_PARTIAL_OUTPUT_ON_ERROR, Json::PARTIAL_OUTPUT_ON_ERROR);
+			$this->assertEquals(JSON_ERROR_RECURSION, JSON_ERROR_RECURSION);
+			$this->assertEquals(JSON_ERROR_INF_OR_NAN, JSON_ERROR_INF_OR_NAN);
+			$this->assertEquals(JSON_ERROR_UNSUPPORTED_TYPE, JSON_ERROR_UNSUPPORTED_TYPE);
+			$this->assertEquals(JSON_PARTIAL_OUTPUT_ON_ERROR, JSON_PARTIAL_OUTPUT_ON_ERROR);
```
Comment (gossi) [inline tests/json/JsonTest.php]: guess, this test is obsolete now ;)

**Review by gossi — COMMENTED** [review]
Really cool. Thanks a ton.

Left one question and I think I found some more thing to remove :)

File: src/file/FileOperationTrait.php (hunk 1 above)
Comment (cristianoc72) [reply src/file/FileOperationTrait.php]: No, it isn't. I was experimenting and I forgot to rollback it.

File: tests/json/JsonTest.php (hunk 2 above)
Comment (cristianoc72) [reply tests/json/JsonTest.php]: :smile: Yes, it is.

**Review by gossi — APPROVED** [review]

**Pull request merged**